src/EditorFeatures/CSharp/CommentSelection/ToggleBlockCommentCommandHandler.cs

```diff
@@ -0,0 +1,428 @@
+    [ContentType(ContentTypeNames.RoslynContentType)]
+    [Name(PredefinedCommandHandlerNames.CommentSelection)]*/
+    internal class ToggleBlockCommentCommandHandler :
+        AbstractCommentSelectionCommandHandler/*,
+        VSCommanding.ICommandHandler<CommentSelectionCommandArgs>*/
+    {
+        [ImportingConstructor]
+        internal ToggleBlockCommentCommandHandler(
+            ITextUndoHistoryRegistry undoHistoryRegistry,
+            IEditorOperationsFactoryService editorOperationsFactoryService) : base(undoHistoryRegistry, editorOperationsFactoryService)
+        { }
+
+        /* TODO - modify once the toggle block comment handler is added.
+        public VSCommanding.CommandState GetCommandState(CommentSelectionCommandArgs args)
+        {
+            return GetCommandState(args.SubjectBuffer);
+        }
+
+        public bool ExecuteCommand(CommentSelectionCommandArgs args, CommandExecutionContext context)
+        {
+            return ExecuteCommand(args.TextView, args.SubjectBuffer, Operation.Toggle, context);
+        }*/
+
+        public override string DisplayName => EditorFeaturesResources.Toggle_Block_Comment;
+
+        protected override string GetTitle(Operation operation)
+        {
+            return EditorFeaturesResources.Toggle_Block_Comment;
+        }
+
+        protected override string GetMessage(Operation operation)
+        {
+            return EditorFeaturesResources.Toggling_block_comment_on_selection;
+        }
+
+        protected override void SetTrackingSpans(ITextView textView, ITextBuffer buffer, List<CommentTrackingSpan> trackingSpans)
+        {
+            var spans = trackingSpans.Select(trackingSpan => trackingSpan.ToSelection(buffer));
+            textView.GetMultiSelectionBroker().SetSelectionRange(spans, spans.Last());
+        }
+
+        internal override void CollectEdits(Document document, ICommentSelectionService service, NormalizedSnapshotSpanCollection selectedSpans,
+            List<TextChange> textChanges, List<CommentTrackingSpan> trackingSpans, Operation operation, CancellationToken cancellationToken)
+        {
+            var experimentationService = document.Project.Solution.Workspace.Services.GetRequiredService<IExperimentationService>();
+            if (!experimentationService.IsExperimentEnabled(WellKnownExperimentNames.RoslynToggleBlockComment))
+            {
+                return;
+            }
+
+            if (selectedSpans.IsEmpty())
+            {
+                return;
+            }
+
+            var getRoot = document.GetSyntaxRootAsync();
+            var commentInfo = service.GetInfoAsync(document, selectedSpans.First().Span.ToTextSpan(), cancellationToken).WaitAndGetResult(cancellationToken);
+            var root = getRoot.WaitAndGetResult(cancellationToken);
+            if (commentInfo.SupportsBlockComment)
+            {
+                ToggleBlockComments(commentInfo, root, selectedSpans, textChanges, trackingSpans);
+            }
```

Comment:
if the language doesn't support block-comments, shouldn't toggle-comments still work?  For example, if i'm in vb, and i select some lines of code, i would expect to eb able to say "toggle" twice and have it comment/uncomment.  Is there a reason we're limiting this only to block comments?

Reply:
This toggles block comments only.  There is a plan for toggling single line comments as well, but not handled in this PR.

Reply:
That seems like a strange omission.  But ok.  Can you add a comment in the code with a link to the plan you're talking about?

Also, given that there's a plan to support single-line-comments, we should def just call this ToggleCommentCommandHandler.  We should then jsut have an 'if' in the code that checks for block commetns and proceeds.  It can have a comment saying that single-line is coming in the future, witha  link to the plan as well.

Reply:
Note that the current proposal for that part isn't fully fleshed out.  Will reference here when it's written down.  However, it would be a *separate* VS command to toggle just single line comments (no effect on block comments).

If we end up being able to re-use pieces of this block comment toggle for single line toggle, I think it makes more sense to change this then, rather than preemptively doing so.

Reply:
>Note that the current proposal for that part isn't fully fleshed out. Will reference here when it's written down. However, it would be a separate VS command to toggle just single line comments (no effect on block comments).

Wacky.  I've left my feedback on the linked design bug giving my thoughts on things.  But i'm ok witht his part of the PR proceeding as is.